### DISCUSSION TITLE
Ho to interpret telegrams?

### DISCUSSION THREAD

**Discussion by hmoslow** [emsesp/EMS-ESP32#1612] [Q&A] [answered]
Hi all,

I have a hybridsystem with an Bosch GC 800IF gas boiler and an Bosch Compress CS 7001i aw heatpump.

The heatpump communicates with the hybridmanager HM200 via canbus. The HM200 submits the data by ems to the controller CW 400.

The delivered telegrams submitted from HM200 to CW400 aren`t not well known by now.

I’m missing the count of starts of the pump, consumed energy to heat water and some more.

I am trying to analyze the telegrams to find the necessary information. But I don't really know anything about it. Where can I find the information for the value type, addition, factor, UOM in the telegrams, for example in these telegrams?

000+19:52:44.498 N 11051: [emsesp] heatpump(0x53) -B-> All(0x00), ?(0x09A6), data: 2B (offset 17)
000+19:49:54.967 N 10262: [emsesp] heatpump(0x53) -B-> All(0x00), ?(0x099D), data: 00 02 00 00 02 00 01 C0 3F 00 3D 03 E8 00 00 00 00
![image](https://github.com/emsesp/EMS-ESP32/assets/158447132/29d40da5-542a-44ad-8600-bdbac1d56b34)
![image](https://github.com/emsesp/EMS-ESP32/assets/158447132/0b1673dc-5b68-4688-b9db-7756c2c06fe4)

Thanks in advance for your reply

**Comment by MichaelDvP** [comment, marked as answer]
> Where can I find the information for the value type, addition, factor, UOM in the telegrams, for example in these telegrams?

Best overview of known telegrams is from [Norberts1](https://github.com/norberts1/hometop_HT3/blob/master/HT3/docu/HT_EMS_Bus_messages.pdf).
In general we can say:
- measurement values are broadcasted periodical 10 sec / 1 min
- settings are only broadcasted after a change
- changing a setting of a device with the UI of thermostat results in a message `thermostat -w-> device` with only this value.
- some devices broadcast fast changing values as single values
- measured temperatures are normally 2 bytes (SHORT) with factor 0.1 (e.g. `01 23` -> 0x0123 -> dez 291 -> 29.1°C)
- air temperature settings are often factor 0.5 as single byte (INT) (e.g. 0x2D -> dez 45 -> 22.5°C)
- water temperature settings are mostly single byte (UINT) (e.g. 0x3C -> 60°C), differential values (hysteresis in Kelvin) are signed (INT).
- percent settings are single byte (UINT) (0x64 -> 100%).
- on/off stati or settings can be single byte with on/off 0xFF/0x00, or 0x01/0x00 or a single bit in a byte together with 7 other stati.
- times and energy is mostly 3 or 4 bytes with or without factor.

For different brands/devices Bosch sometimes use different expressions for the same value, Maybe changing developers or they like to make reverse engineering difficult.

If you search a setting, log the telegrams for the device (log all or `watch <device-id>`) and change the setting on the thermostat to different states/values, Then search these values in the log.
If you search for a measurement, log the device and view the value on the thermostat and wait for changes, note old/new values and time. Then check the log for this time (or 10 sec / 1 min later) and value withing a telegram. Best to have more changes/values.
